cob_hardware_emulation/scripts/emulation_base.py

```diff
@@ -158,6 +177,49 @@ def timer_cb(self, event):
 
         self.br.sendTransform(transforms)
 
+    def move_base_cb(self, goal):
+        pwcs = PoseWithCovarianceStamped()
+        pwcs.header = goal.target_pose.header
+        pwcs.pose.pose = goal.target_pose.pose
+        if self.move_base_mode == "beam":
+            rospy.loginfo("move_base: beaming robot to new goal")
+            self.initalpose_callback(pwcs)
+        elif self.move_base_mode == "linear_nav":
+            move_base_linear_action_name = "/move_base_linear"
+            ac_move_base_linear = actionlib.SimpleActionClient(move_base_linear_action_name, MoveBaseAction)
+            rospy.loginfo("Waiting for ActionServer: %s", move_base_linear_action_name)
+            if not ac_move_base_linear.wait_for_server(rospy.Duration(1)):
+                rospy.logerr("Emulator move_base failed because move_base_linear action server is not available")
+                self.as_move_base.set_aborted()
+                return
+            rospy.loginfo("send goal to %s", move_base_linear_action_name)
+            ac_move_base_linear.send_goal(goal)
+            ac_move_base_linear.wait_for_result()
+            ac_move_base_linear_status = ac_move_base_linear.get_state()
+            ac_move_base_linear_result = ac_move_base_linear.get_result()
+            if ac_move_base_linear_status != GoalStatus.SUCCEEDED:
+                rospy.logerr("Emulator move_base failed because move_base_linear failed")
+                self.as_move_base.set_aborted()
+                return
+        else:
+            rospy.logerr("Invalid move_base_action_mode")
+            self.as_move_base.set_aborted()
+            return
+        rospy.loginfo("Emulator move_base succeeded")
+        self.as_move_base.set_succeeded(MoveBaseResult())
+
+    def move_base_simple_callback(self, msg):
+        goal = MoveBaseGoal()
+        goal.target_pose = msg
+        ac_move_base = actionlib.SimpleActionClient(self.move_base_action_name, MoveBaseAction)
+        rospy.loginfo("Waiting for ActionServer: %s", self.move_base_action_name)
+        if not ac_move_base.wait_for_server(rospy.Duration(1)):
+            rospy.logerr("Emulator move_base simple failed because move_base action server is not available")
+            return
+        rospy.loginfo("send goal to %s", self.move_base_action_name)
+        ac_move_base.send_goal(goal)
```

Comment:
This is calling the own action_server. Can't you just call the internal function `move_base_cb` instead?

Reply:
yes, I did that first. But then I couldn't get the action handle correctly. So the `move_base_cb` complained about setting a result (e.g. succeeded in Line 209) for an invalid action handle.

Reply:
ah ok. sure, if no action is active the action methods (set_aborted, set_succeeded...) will fail. We could separate the action logic from the application logic into separate methods. But I think that is fine for now...